packages/bitcore-node/src/services/pruning.ts

```diff
@@ -34,45 +36,101 @@ export class PruningService {
   async detectAndClear() {
     for (let chainNetwork of Config.chainNetworks()) {
       const { chain, network } = chainNetwork;
-      await this.detectInvalidMempoolCoins(chain, network);
-      const invalids = this.detectInvalidCoins(chain, network);
-      for await (const invalidCoins of invalids) {
-        if (this.stopping) {
-          return;
-        }
-        const txids = invalidCoins.map(c => c.mintTxid);
-        await this.clearInvalid(txids);
-      }
+      await this.processOldMempoolTxs(chain, network, MEMPOOL_AGE);
+      await this.processAllInvalidTxs(chain, network);
     }
   }
 
-  async detectInvalidMempoolCoins(chain, network) {
-    const txs = await this.transactionModel.collection.find({ chain, network, blockHeight: -3 }).toArray();
-    for (let groups of partition(txs, 1000)) {
-      const updated = await this.coinModel.collection.updateMany(
-        { chain, network, mintTxid: { $in: groups.map(tx => tx.txid) }, mintHeight: -1 },
-        { $set: { mintHeight: -3 } }
-      );
-      logger.info('Mempool coins updated', updated);
-    }
+  async processOldMempoolTxs(chain: string, network: string, days: number) {
+    const ONE_HOUR = 60 * 60 * 1000;
+    const ONE_DAY = 24 * ONE_HOUR;
+    const oldTime = new Date(Date.now() - days * ONE_DAY);
+    const count = await this.transactionModel.collection.countDocuments({
+      chain,
+      network,
+      blockHeight: -1,
+      blockTimeNormalized: { $lt: oldTime }
+    });
+    logger.info(`Found ${count} outdated ${chain} ${network} mempool txs`);
+    await new Promise((resolve, reject) => {
+      this.transactionModel.collection
+        .find({ chain, network, blockHeight: -1, blockTimeNormalized: { $lt: oldTime } })
+        .pipe(
+          new Transform({
+            objectMode: true,
+            transform: async (data: any, _, cb) => {
+              if (this.stopping) {
+                return cb(new Error('Stopping'));
+              }
+              const tx = data as ITransaction;
+              const outputs = await this.transactionModel.findAllRelatedOutputs(tx.txid);
+              const invalid = outputs.find(c => c.mintHeight >= 0 || c.spentHeight >= 0);
```

Comment:
Is this something that could happen? Should we log this issue?

Reply:
I think it should crash if that happens, and log, I'll add an error there